### Project 2. IncomeQalification


In [2]:
#inporting required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [3]:
# importing the datasets for both train and test
df_train=pd.read_csv('/content/train.csv')
df_test=pd.read_csv('/content/test.csv')


In [6]:
df_train.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  ...  SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
0     0  ...          100    1849               1        100             0   
1     0  ...          144    4489               1        144             0   
2     0  ...          121    8464               1          0             0   
3     0  ...           81     289              16        121             4   
4     0  ...          121    1369              16        121             4   

   SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
0         1.000000            0.0      100.0   1849       4  
1         1.000000           64.0      144.0   4489       4  
2         0.250000           64.0      121.0   8464       4  
3         1.777778            1.0      121.0    289       4  
4         1.777778            1.0      121.0   1369       4  

[5 rows x 143 columns]

In [4]:
#Identifying the data types of training data
dtypes=df_train.dtypes.to_dict()
dtypes

{'Id': dtype('O'),
 'SQBage': dtype('int64'),
 'SQBdependency': dtype('float64'),
 'SQBedjefe': dtype('int64'),
 'SQBescolari': dtype('int64'),
 'SQBhogar_nin': dtype('int64'),
 'SQBhogar_total': dtype('int64'),
 'SQBmeaned': dtype('float64'),
 'SQBovercrowding': dtype('float64'),
 'Target': dtype('int64'),
 'abastaguadentro': dtype('int64'),
 'abastaguafuera': dtype('int64'),
 'abastaguano': dtype('int64'),
 'age': dtype('int64'),
 'agesq': dtype('int64'),
 'area1': dtype('int64'),
 'area2': dtype('int64'),
 'bedrooms': dtype('int64'),
 'cielorazo': dtype('int64'),
 'computer': dtype('int64'),
 'coopele': dtype('int64'),
 'dependency': dtype('O'),
 'dis': dtype('int64'),
 'edjefa': dtype('O'),
 'edjefe': dtype('O'),
 'elimbasu1': dtype('int64'),
 'elimbasu2': dtype('int64'),
 'elimbasu3': dtype('int64'),
 'elimbasu4': dtype('int64'),
 'elimbasu5': dtype('int64'),
 'elimbasu6': dtype('int64'),
 'energcocinar1': dtype('int64'),
 'energcocinar2': dtype('int64'),
 'energcocinar3': dtype('

We have data of three data types i.e. float64,int64 and object

In [5]:
#info
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 143 entries, Id to Target
dtypes: float64(8), int64(130), object(5)
memory usage: 10.4+ MB


In [7]:
#Dropping the id columns as they arent required
df_train=df_train.drop(["Id","idhogar"],axis=1)

In [8]:
df_test=df_test.drop(["Id", "idhogar"],axis=1)

In [9]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 141 entries, v2a1 to Target
dtypes: float64(8), int64(130), object(3)
memory usage: 10.3+ MB


In [10]:
#  getting all object dtypes from the df_train.columns
for i in df_train.columns:
  a=df_train[i].dtype
  if a=='object':
    print(i)

dependency
edjefe
edjefa


In [11]:
#Creating a function for mapping the value yes with 1 and no with 0
def map(i):
  if i=='yes':
    return(float(1))
  elif i=='no':
    return(float(0)) 
  else:
    return(float(i))  

In [12]:
# applying function map on dependency , edjefe and edjefa
df_train['dependency']=df_train['dependency'].apply(map)


In [13]:
df_train['edjefe']=df_train['edjefe'].apply(map)
df_train['edjefa']=df_train['edjefa'].apply(map)

In [14]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 141 entries, v2a1 to Target
dtypes: float64(11), int64(130)
memory usage: 10.3 MB


### Identifying the variance of each column

In [15]:
df_train.var()

v2a1               2.263735e+10
hacdor             3.664046e-02
rooms              2.156143e+00
hacapo             2.309080e-02
v14a               5.204941e-03
                       ...     
SQBovercrowding    1.705316e+01
SQBdependency      1.565459e+02
SQBmeaned          8.745409e+03
agesq              3.031767e+06
Target             1.019221e+00
Length: 141, dtype: float64

In [16]:
#columns with zero variace
for i in df_train:
  v=df_train[i].var()
  if v== 0:
    print(i)

elimbasu5


In [17]:
# droping columns elimbasu5 for both train and test datasets
df_train=df_train.drop("elimbasu5",axis=1)
df_test=df_test.drop("elimbasu5",axis=1)

In [18]:
#total person in household
df_train["r4t3"].value_counts()

4     2447
3     2260
5     1598
2     1355
6      824
1      378
7      368
8       96
9       90
10      60
11      44
12      24
13      13
Name: r4t3, dtype: int64

In [19]:
df_train=df_train.drop("r4t3",axis=1)
df_test=df_test.drop("r4t3",axis=1)

Check if there is a house without a family head.

In [20]:
df_train['parentesco1'].value_counts()

0    6584
1    2973
Name: parentesco1, dtype: int64

In [21]:
pd.crosstab(df_train['edjefa'],df_train['edjefe'])

edjefe  0.0   1.0   2.0   3.0   4.0   5.0   6.0   7.0   8.0   9.0   ...  12.0  \
edjefa                                                              ...         
0.0      435   123   194   307   137   222  1845   234   257   486  ...   113   
1.0       69     0     0     0     0     0     0     0     0     0  ...     0   
2.0       84     0     0     0     0     0     0     0     0     0  ...     0   
3.0      152     0     0     0     0     0     0     0     0     0  ...     0   
4.0      136     0     0     0     0     0     0     0     0     0  ...     0   
5.0      176     0     0     0     0     0     0     0     0     0  ...     0   
6.0      947     0     0     0     0     0     0     0     0     0  ...     0   
7.0      179     0     0     0     0     0     0     0     0     0  ...     0   
8.0      217     0     0     0     0     0     0     0     0     0  ...     0   
9.0      237     0     0     0     0     0     0     0     0     0  ...     0   
10.0      96     0     0     0     0     0     0     0     0     0  ...     0   
11.0     399     0     0     0     0     0     0     0     0     0  ...     0   
12.0      72     0     0     0     0     0     0     0     0     0  ...     0   
13.0      52     0     0     0     0     0     0     0     0     0  ...     0   
14.0     120     0     0     0     0     0     0     0     0     0  ...     0   
15.0     188     0     0     0     0     0     0     0     0     0  ...     0   
16.0     113     0     0     0     0     0     0     0     0     0  ...     0   
17.0      76     0     0     0     0     0     0     0     0     0  ...     0   
18.0       3     0     0     0     0     0     0     0     0     0  ...     0   
19.0       4     0     0     0     0     0     0     0     0     0  ...     0   
20.0       2     0     0     0     0     0     0     0     0     0  ...     0   
21.0       5     0     0     0     0     0     0     0     0     0  ...     0   

edjefe  13.0  14.0  15.0  16.0  17.0  18.0  19.0  20.0  21.0  
edjefa                                                        
0.0      103   208   285   134   202    19    14     7    43  
1.0        0     0     0     0     0     0     0     0     0  
2.0        0     0     0     0     0     0     0     0     0  
3.0        0     0     0     0     0     0     0     0     0  
4.0        0     0     0     0     0     0     0     0     0  
5.0        0     0     0     0     0     0     0     0     0  
6.0        0     0     0     0     0     0     0     0     0  
7.0        0     0     0     0     0     0     0     0     0  
8.0        0     0     0     0     0     0     0     0     0  
9.0        0     0     0     0     0     0     0     0     0  
10.0       0     0     0     0     0     0     0     0     0  
11.0       0     0     0     0     0     0     0     0     0  
12.0       0     0     0     0     0     0     0     0     0  
13.0       0     0     0     0     0     0     0     0     0  
14.0       0     0     0     0     0     0     0     0     0  
15.0       0     0     0     0     0     0     0     0     0  
16.0       0     0     0     0     0     0     0     0     0  
17.0       0     0     0     0     0     0     0     0     0  
18.0       0     0     0     0     0     0     0     0     0  
19.0       0     0     0     0     0     0     0     0     0  
20.0       0     0     0     0     0     0     0     0     0  
21.0       0     0     0     0     0     0     0     0     0  

[22 rows x 22 columns]

 435 families with no family head.

### finding the null values

In [22]:
for i in df_train:
  v=df_train[i].isnull().sum()
  if v>0:
    print("Column Name: ",i, ", Null Values for this column: ",v,", Percentage of Null values: ",df_train[i].isnull().sum()/len(df_train)*100)

Column Name:  v2a1 , Null Values for this column:  6860 , Percentage of Null values:  71.7798472323951
Column Name:  v18q1 , Null Values for this column:  7342 , Percentage of Null values:  76.82327090091033
Column Name:  rez_esc , Null Values for this column:  7928 , Percentage of Null values:  82.95490216595167
Column Name:  meaneduc , Null Values for this column:  5 , Percentage of Null values:  0.05231767290990897
Column Name:  SQBmeaned , Null Values for this column:  5 , Percentage of Null values:  0.05231767290990897


In [23]:
df_train["v2a1"].describe()

count    2.697000e+03
mean     1.652316e+05
std      1.504571e+05
min      0.000000e+00
25%      8.000000e+04
50%      1.300000e+05
75%      2.000000e+05
max      2.353477e+06
Name: v2a1, dtype: float64

  'v2a1', 'v18q1', 'rez_esc' have more than 50% null values, because for v18q1, there are families with their own house so they won't pay rent in that case it should be 0

In [24]:
#filling the missing value fo both train and test datasets using fillna()
df_train['v2a1']=df_train['v2a1'].fillna(0)
df_train['v18q1']=df_train['v18q1'].fillna(0)
df_test['v2a1']=df_test['v2a1'].fillna(0)
df_test['v18q1']=df_test['v18q1'].fillna(0)

In [25]:
df_train=df_train.drop(['tipovivi3', 'v18q','rez_esc'],axis=1)
df_test=df_test.drop(['tipovivi3', 'v18q','rez_esc'],axis=1)

In [26]:
# fillna() using np.mean on train and test 
df_train['meaneduc']=df_train['meaneduc'].fillna(np.mean(df_train['meaneduc']))
df_train['SQBmeaned']=df_train['SQBmeaned'].fillna(np.mean(df_train['SQBmeaned']))
df_test['meaneduc']=df_test['meaneduc'].fillna(np.mean(df_test['meaneduc']))
df_test['SQBmeaned']=df_test['SQBmeaned'].fillna(np.mean(df_test['SQBmeaned']))

## Set the poverty level of the members and the head of the house within a family.

as peple below poverty level can be person paying less rent, do not own house and also depend on whether house is in rural area or urban area 

In [27]:
Poverty_level=df_train[df_train['v2a1'] !=0]

In [28]:
Poverty_level.shape

(2668, 136)

In [29]:
# groping poverty level with area1 i,e zonal urbana and v2a1--> monthaly rented
poverty_level=Poverty_level.groupby('area1')['v2a1'].apply(np.median)

In [30]:
poverty_level

area1
0     80000.0
1    140000.0
Name: v2a1, dtype: float64

In [31]:
def povert(x):
    if x<8000:
        return('Below poverty level')
    
    elif x>140000:
        return('Above poverty level')
    elif x<140000:
        return('Below poverty level: Ur-ban ; Above poverty level : Rural ')

In [32]:
c=Poverty_level['v2a1'].apply(povert)

In [33]:
c.shape

(2668,)

In [34]:
pd.crosstab(c,Poverty_level['area1'])


area1                                                 0     1
v2a1                                                         
Above poverty level                                 139  1103
Below poverty level: Ur-ban ; Above poverty lev...  306  1081

assigning x and y

In [35]:
x=df_train.drop('Target',axis=1)
y=df_train.Target

In [36]:
X_data_col=x.columns


## Applying Standard Scalling to dataset

In [37]:
from sklearn.preprocessing import StandardScaler
SS=StandardScaler()
X_data_1=SS.fit_transform(x)
X_data_1=pd.DataFrame(X_data_1,columns=X_data_col)

In [39]:
#applying train_test_split on scaled data
from sklearn.model_selection import train_test_split

In [40]:
x_train,x_test,y_train,y_test=train_test_split(X_data_1,y,test_size=0.25,stratify=y,random_state=0)

### predicting the accuracy using random forest classifier

In [41]:
from sklearn.ensemble import RandomForestClassifier

In [42]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

In [43]:
rfc=RandomForestClassifier(random_state=0)
parameters={'n_estimators':[10,50,100,300],'max_depth':[3,5,10,15]}
grid=zip([rfc],[parameters])

best_=None

In [44]:
for i, j in grid:
    a=GridSearchCV(i,param_grid=j,cv=3,n_jobs=1)
    a.fit(x_train,y_train)
    if best_ is None:
        best_=a
    elif a.best_score_>best_.best_score_:
        best_=a
        

In [45]:
print ("Best CV Score",best_.best_score_)
print ("Model Parameters",best_.best_params_)
print("Best Estimator",best_.best_estimator_)

Best CV Score 0.8507046183898423
Model Parameters {'max_depth': 15, 'n_estimators': 300}
Best Estimator RandomForestClassifier(max_depth=15, n_estimators=300, random_state=0)


In [48]:
#applying best estimators on random forest classifier
RFC=best_.best_estimator_
Model=RFC.fit(x_train,y_train)
pred=Model.predict(x_test)

In [47]:
print('Model Score of train data : {}'.format(Model.score(x_train,y_train)))
print('Model Score of test data : {}'.format(Model.score(x_test,y_test)))

Model Score of train data : 0.9831170643225896
Model Score of test data : 0.8824267782426778


In [49]:
Important_features=pd.DataFrame(Model.feature_importances_,X_data_col,columns=['feature_importance'])


In [50]:
Top50Features=Important_features.sort_values(by='feature_importance',ascending=False).head(50).index

In [51]:
Top50Features

Index(['SQBmeaned', 'meaneduc', 'SQBdependency', 'dependency', 'overcrowding',
       'SQBovercrowding', 'qmobilephone', 'SQBhogar_nin', 'SQBedjefe',
       'edjefe', 'hogar_nin', 'rooms', 'cielorazo', 'r4t1', 'v2a1', 'edjefa',
       'agesq', 'r4m3', 'r4h2', 'SQBage', 'age', 'escolari', 'r4t2', 'r4h3',
       'hogar_adul', 'SQBescolari', 'eviv3', 'bedrooms', 'r4m1', 'epared3',
       'r4m2', 'tamviv', 'paredblolad', 'v18q1', 'SQBhogar_total', 'tamhog',
       'hhsize', 'hogar_total', 'pisomoscer', 'etecho3', 'r4h1', 'lugar1',
       'eviv2', 'tipovivi1', 'energcocinar2', 'energcocinar3', 'epared2',
       'television', 'area2', 'area1'],
      dtype='object')

In [52]:
for i in Top50Features:
    if i not in X_data_col:
        print(i)

In [53]:
X_data_Top50=x[Top50Features]

In [54]:
# applying train_test_split on top 50 features to reduce overfitting
x_train,x_test,y_train,y_test=train_test_split(X_data_Top50,y,test_size=0.25,stratify=y,random_state=0)

In [55]:
# fitting x_train, y_train on random forest classifier
Model_1=RFC.fit(x_train,y_train)
pred=Model_1.predict(x_test)

In [57]:
# inporting confusion_matrix, f1_score and accuracy_score
from sklearn.metrics import confusion_matrix,f1_score,accuracy_score

In [58]:
confusion_matrix(y_test,pred)

array([[ 143,   17,    0,   29],
       [   8,  324,    4,   63],
       [   1,   12,  214,   75],
       [   2,   10,    3, 1485]])

In [59]:
f1_score(y_test,pred,average='weighted')

0.9026906492316511

In [60]:
accuracy_score(y_test,pred)

0.906276150627615

In [61]:
df_test['dependency']=df_test['dependency'].apply(map)

In [62]:
df_test['edjefe']=df_test['edjefe'].apply(map)
df_test['edjefa']=df_test['edjefa'].apply(map)

In [63]:
test_data=df_test[Top50Features]

In [64]:
test_data.isna().sum().value_counts()

0    50
dtype: int64

In [66]:
# scaling test data by standard scaling
Test_data_1=SS.fit_transform(test_data)


In [67]:
X_data_1=pd.DataFrame(Test_data_1)

In [68]:
test_prediction=Model_1.predict(test_data)

In [69]:
test_prediction

array([4, 4, 4, ..., 4, 4, 4])

using random forest classifier test data is predicted with the accuracy of **90**%